server/coordinator/procedure/drop_partition_table.go

```diff
@@ -0,0 +1,421 @@
+		cancelEventWithLog(event, ErrDropTableResult, fmt.Sprintf("legnth of shardVersionResult need >=1, current is %d", len(dropTableRet.ShardVersionUpdate)))
+		return
+	}
+
+	request.versions = dropTableRet.ShardVersionUpdate
+	request.table = table
+
+	// Drop table in the first shard.
+	if err := dispatchDropTable(event, table, dropTableRet.ShardVersionUpdate[0]); err != nil {
+		cancelEventWithLog(event, err, fmt.Sprintf("drop table, table:%s", table.Name))
+		return
+	}
+}
+
+// 3. Close partition table in target node.
+func closePartitionTableCallback(event *fsm.Event) {
+	request, err := getRequestFromEvent[*dropPartitionTableCallbackRequest](event)
+	if err != nil {
+		cancelEventWithLog(event, err, "get request from event")
+		return
+	}
+
+	var shardNodes []storage.ShardNode
+	for _, version := range request.versions {
+		ret, err := request.cluster.GetShardNodesByShardID(version.ShardID)
+		if err != nil {
+			cancelEventWithLog(event, err, "get shard nodes by shard id")
+			return
+		}
+		shardNodes = append(shardNodes, ret...)
+	}
+
+	tableInfo := cluster.TableInfo{
+		ID:         request.table.ID,
+		Name:       request.table.Name,
+		SchemaID:   request.table.SchemaID,
+		SchemaName: request.sourceReq.GetSchemaName(),
+	}
+
+	for _, shardNode := range shardNodes {
+		// Close partition table shard.
+		if err = request.dispatch.CloseTableOnShard(request.ctx, shardNode.NodeName, eventdispatch.CloseTableOnShardRequest{
+			UpdateShardInfo: eventdispatch.UpdateShardInfo{},
+			TableInfo:       tableInfo,
+		}); err != nil {
+			cancelEventWithLog(event, err, "close shard")
+			return
+		}
+	}
+}
+
+func finishDropPartitionTableCallback(event *fsm.Event) {
+	request, err := getRequestFromEvent[*dropPartitionTableCallbackRequest](event)
+	if err != nil {
+		cancelEventWithLog(event, err, "get request from event")
+		return
+	}
+	log.Info("drop partition table finish")
+
+	tableInfo := cluster.TableInfo{
+		ID:         request.table.ID,
+		Name:       request.table.Name,
+		SchemaID:   request.table.SchemaID,
+		SchemaName: request.sourceReq.GetSchemaName(),
+	}
+
+	if err = request.onSucceeded(tableInfo); err != nil {
+		cancelEventWithLog(event, err, "drop partition table on succeeded")
+		return
+	}
+}
+
+func dropTableMetaData(event *fsm.Event, tableName string) (storage.Table, cluster.DropTableResult, error) {
+	request, err := getRequestFromEvent[*dropPartitionTableCallbackRequest](event)
+	if err != nil {
+		return storage.Table{}, cluster.DropTableResult{}, errors.WithMessage(err, "get request from event")
+	}
+
+	table, exists, err := request.cluster.GetTable(request.schemaName(), tableName)
+	if err != nil {
+		return storage.Table{}, cluster.DropTableResult{}, errors.WithMessage(err, "cluster get table")
+	}
+	if !exists {
+		log.Warn("drop non-existing table", zap.String("schema", request.schemaName()), zap.String("table", tableName))
+		return storage.Table{}, cluster.DropTableResult{}, ErrTableNotExists
```

Comment:
If return `ErrTableNotExists` when table not exists, then `dropPartitionTable` and `dropDataTable` are not idempotent functions.